Rust/His: Initial work to generalize scan definitions

beaubelgrave opened the pull request:
Today we have a very fast, but labor intensive, way to add signatures. If we need to add more or have a way to gracefully have some but not others for performance, that is currently very hard to achieve.

Add a generic ScanDefinition struct that can take either a 3 or 4 byte signature along with the character to check for in every 16 bytes. Right now keep the enum and static checking, however, add a closure slot in the definition to indicate the final match behavior in the future.

Each ScanDefinition includes how far back from the siganture and how far forward from the signature to get for a final pass. The underlying structure handles the UTF8 vs UTF16 differences. Callers just need to define UTF8 once, UTF16 LE and BE are auto-calculated.

Currently the performance is around 60-80% slower than hand curated switch/jump tables. However, it is still quite a bit faster than any alternative methods.

cfaucon reviewed (commented):
Did run all the tests for this package when targeting net45 to ensure they all pass?

michaelcfanning commented:
We don't want to update this dependency to support and out-of-support .net. Instead, I suggest that you simply rebuild this code for that dependency in your own local context. You can use a sub-module or simply create a copy of the code. If you figure out why this compilation caused test problems for you in your separate utilization, I would be greatly to hear why. :)

---
In reply to: [896946542](https://github.com/microsoft/security-utilities/pull/21#issuecomment-896946542) [](http://example.com/codeflow?ancestors=896946542,896946542)

beaubelgrave committed:
Rust/His: Initial work to generalize scan definitions

Today we have a very fast, but labor intensive, way to add signatures.
If we need to add more or have a way to gracefully have some but not
others for performance, that is currently very hard to achieve.

Add a generic ScanDefinition struct that can take either a 3 or 4 byte
signature along with the character to check for in every 16 bytes. Right
now keep the enum and static checking, however, add a closure slot in
the definition to indicate the final match behavior in the future.

Each ScanDefinition includes how far back from the siganture and how far
forward from the signature to get for a final pass. The underlying
structure handles the UTF8 vs UTF16 differences. Callers just need to
define UTF8 once, UTF16 LE and BE are auto-calculated.

Currently the performance is around 60-80% slower than hand curated
switch/jump tables. However, it is still quite a bit faster than any
alternative methods.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

michaelcfanning reviewed (approved):
:shipit:
